File: src/util/project.rs
use anyhow::{anyhow, bail};
use log::{debug, trace};
use serde::Serialize;
use std::{borrow::Cow, path::PathBuf, str::FromStr};
use url::Url;

use crate::util::nix::{self, EvalResult};

#[derive(Debug)]
pub enum Source {
    Path { path: PathBuf },
    Git { info: GitInfo, path: PathBuf },
    Sourcehut { info: GitXInfo, path: PathBuf },
    Tarball { url: String, path: PathBuf },
}

impl Source {
    pub fn get_path(self) -> PathBuf {
        match self {
            Source::Path { path } => path,
            Source::Git { info: _, path } => path,
            Source::Sourcehut { info: _, path } => path,
            Source::Tarball { url: _, path } => path,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct GitInfo {
    pub url: String,
    pub rev: Option<String>,
    pub r#ref: Option<String>,
    pub dir: Option<String>,
    pub submodules: bool,
}

#[derive(Debug)]
pub struct GitXInfo {
    pub owner: String,
    pub repo: String,
    pub rev: Option<String>,
    pub r#ref: Option<String>,
    pub dir: Option<String>,
    pub host: String,
    pub submodules: bool,
}

impl From<GitXInfo> for GitInfo {
    fn from(value: GitXInfo) -> Self {
        GitInfo {
            url: format!("git@{}:{}/{}.git", value.host, value.owner, value.repo),
            rev: value.rev,
            r#ref: value.r#ref,
            dir: value.dir,
            submodules: value.submodules,
        }
    }
}

fn remove_filename_from_path(mut path: PathBuf) -> PathBuf {
    if path.is_file() {
        trace!("Splicing off {:?}", path.file_name());
        path.pop();
    }
    path
}

async fn resolve_git(info: GitInfo) -> anyhow::Result<Source> {
    trace!("Resolving for {info:?}");
    let code = format!(
        "
		let
			info = builtins.fromJSON ''{}'';
		in
			builtins.fetchGit (
				{{ url = info.url; }}
				// (if info.rev != null then {{ rev = info.rev; }} else {{}})
				// (if info.ref != null then {{ ref = info.ref; }} else {{}})
				// (if info.submodules != null then {{ submodules = info.submodules; }} else {{}})
			)
	",
        serde_json::to_string(&info).unwrap()
    );

    let root = nix::evaluate(
        &code,
        nix::EvalOpts {
            impure: true,
            json: true,
        },
    )
    .await;

    let root_path = match root {
        Ok(EvalResult::Json(res)) => res.as_str().unwrap().to_string(),
        Ok(EvalResult::Raw(_)) => {
            bail!("Got raw, expected JSON");
        }
        _ => {
            bail!("{}", root.unwrap_err());
        }
    };

    let store_path = nix::realise(&PathBuf::from(root_path)).await;

    let Ok(paths) = store_path else {
        bail!("{}", store_path.unwrap_err());
    };

    let mut final_path = paths[0].clone();

    if info.dir.is_some() {
        final_path.push(info.dir.clone().unwrap())
    }

    return Ok(Source::Git {
        info,
        path: final_path,
    });
}

async fn resolve_tar(url: &str) -> anyhow::Result<Source> {
    let code = format!(
        "
		builtins.fetchTarball {{
			url = \"{url}\";
		}}
	"
    );

    let root = nix::evaluate(
        &code.trim(),
        nix::EvalOpts {
            impure: true,
            json: true,
        },
    )
    .await;

    let root_path = match root {
        Ok(EvalResult::Json(res)) => res.as_str().unwrap().to_string(),
        Ok(EvalResult::Raw(_)) => {
            bail!("Got raw, expected JSON");
        }
        _ => {
            bail!("{}", root.unwrap_err());
        }
    };

    let store_path = nix::realise(&PathBuf::from(root_path)).await;

    let Ok(paths) = store_path else {
        bail!("{}", store_path.unwrap_err());
    };

    return Ok(Source::Tarball {
        url: url.to_string(),
        path: paths[0].clone(),
    });
}

pub async fn resolve(uri: &str) -> anyhow::Result<Source> {
    trace!("Trying {uri} as local path");
    if uri.starts_with(".") || uri.starts_with("/") || uri.starts_with("~") {
        if let Ok(real_path) = {
            let Ok(path) = PathBuf::from_str(uri);
            path.canonicalize()
        } {
            debug!("Found path {}", real_path.display());
            return Ok(Source::Path {
                path: remove_filename_from_path(real_path),
            });
        } else {
            bail!("Could not find path {uri}");
        }
    }
    if uri.starts_with("path:") {
        if let Ok(real_path) = {
            let Ok(path) = PathBuf::from_str(&uri[4..]);
            path.canonicalize()
        } {
            debug!("Found path {}", real_path.display());
            return Ok(Source::Path {
                path: remove_filename_from_path(real_path),
            });
        } else {
            bail!("Could not find path {}", &uri[4..])
        };
    }

    trace!("Trying as URL");

    if uri.starts_with("git:") {
        trace!("matched as git");
        let url = Url::parse(uri).unwrap();
        let qps = url.query_pairs();
        let info = GitInfo {
            url: url.path().to_string(),
            rev: qps
                .clone()
                .find(|(k, _)| k == "rev")
                .map(|(_, v)| v.to_string()),
            r#ref: qps
                .clone()
                .find(|(k, _)| k == "ref")
                .map(|(_, v)| v.to_string()),
            dir: qps
                .clone()
                .find(|(k, _)| k == "dir")
                .map(|(_, v)| v.to_string()),
            submodules: qps
                .clone()
                .find(|(k, _)| k == "submodules")
                .unwrap_or(("".into(), "false".into()))
                .1
                .to_string()
                == "true",
        };
        return resolve_git(info).await;
    } else if uri.starts_with("github:") {
        trace!("matched as github");
        let url = Url::parse(&format!("github://{}", &uri[7..])).unwrap();
        let mut parsed = url
            .path_segments()
            .ok_or_else(|| anyhow!("cannot be base"))?;
        let owner = url.host().unwrap().to_string();
        let repo = parsed
            .next()
            .ok_or_else(|| anyhow!("could not get repo"))?
            .to_string();

        let qps = url.query_pairs();

        let info = GitXInfo {
            owner,
            repo,
            r#ref: qps
                .clone()
                .find(|(k, _)| k == "ref")
                .map(|(_, v)| v.to_string()),
            rev: qps
                .clone()
                .find(|(k, _)| k == "rev")
                .map(|(_, v)| v.to_string()),
            dir: qps
                .clone()
                .find(|(k, _)| k == "dir")
                .map(|(_, v)| v.to_string()),
            host: qps
                .clone()
                .find(|(k, _)| k == "host")
                .or(Some((Cow::from(""), Cow::from("github.com"))))
                .unwrap()
                .1
                .to_string(),
            submodules: qps
                .clone()
                .find(|(k, _)| k == "submodules")
                .unwrap_or(("".into(), "false".into()))
                .1
                .to_string()
                == "true",
        };
        return resolve_git(info.into()).await;
    } else if uri.starts_with("gitlab:") {
        trace!("matched as gitlab");
        let url = Url::parse(&format!("gitlab://{}", &uri[7..])).unwrap();
        let mut parsed = url
            .path_segments()
            .ok_or_else(|| anyhow!("cannot be base"))?;
        let owner = url.host().unwrap().to_string();
        let repo = parsed
            .next()
            .ok_or_else(|| anyhow!("could not get repo"))?
            .to_string();

        let qps = url.query_pairs();

        let info = GitXInfo {
            owner,
            repo,
            r#ref: qps
                .clone()
                .find(|(k, _)| k == "ref")
                .map(|(_, v)| v.to_string()),
            rev: qps
                .clone()
                .find(|(k, _)| k == "rev")
                .map(|(_, v)| v.to_string()),
            dir: qps
                .clone()
                .find(|(k, _)| k == "dir")
                .map(|(_, v)| v.to_string()),
            host: qps
                .clone()
                .find(|(k, _)| k == "host")
                .or(Some((Cow::from(""), Cow::from("gitlab.com"))))
                .unwrap()
                .1
                .to_string(),
            submodules: qps
                .clone()
                .find(|(k, _)| k == "submodules")
                .unwrap_or(("".into(), "false".into()))
                .1
                .to_string()
                == "true",
        };

        return resolve_git(info.into()).await;
    } else if uri.starts_with("tarball:") {
        trace!("matched as tarball");
        let mut minus_tar = uri[8..].to_string();
        if !minus_tar.starts_with("http://") && !minus_tar.starts_with("https://") {
            minus_tar = format!("http://{minus_tar}");
        }
        return resolve_tar(&minus_tar).await;
    } else if uri.starts_with("http://") || uri.starts_with("https://") {
        trace!("matched as http(s)");
        return resolve_tar(uri).await;
    } else {
        bail!("Could not parse URL Scheme for {uri}")
    }
}

File: src/util/project.ts
import fs from "node:fs/promises";
import path from "node:path";
import process from "node:process";

import log from "~/util/log";
import exec from "~/util/exec";
import * as nix from "~/util/nix";

export const exists = async (path: string) => {
	try {
		await fs.access(path);
		return true;
	} catch (error) {
		return false;
	}
};

export enum Source {
	Path = "path",
	Git = "git",
	Github = "github",
	Gitlab = "gitlab",
	Sourcehut = "sourcehut",
	Tarball = "tarball",
}

export const getProjectRoot = async (base: string) => {
	let absolute = path.isAbsolute(base) ? base : path.resolve(base);
	let root = absolute;

	while (true) {
		if (root === "/") {
			log.fatal(`No Nilla project found at or above "${absolute}".`);
			process.exit(1);
		}

		try {
			await exists(path.join(root, "nilla.nix"));
			break;
		} catch (error) {
			root = path.dirname(root);
		}
	}

	return root;
};

export const isUrl = (uri: string) => {
	try {
		new URL(uri);
		return true;
	} catch (error) {
		return false;
	}
};

export const getRawUrl = (uri: string): URL => {
	try {
		return new URL(uri);
	} catch (error) {
		log.fatal(`Invalid URI: ${uri}`);
		process.exit(1);
	}
};

export const resolve = async (uri: string) => {
	if (uri.startsWith(".") || uri.startsWith("/")) {
		const root = await getProjectRoot(uri);

		return {
			source: Source.Path,
			path: root,
		};
	}

	if (uri.startsWith("path:")) {
		const root = uri.slice("path:".length);

		if (!(await exists(root))) {
			log.fatal(`Path not found: ${root}`);
			process.exit(1);
		}

		return {
			source: Source.Path,
			path: root,
		};
	}

	const url = getRawUrl(uri);

	switch (url.protocol) {
		case "git:": {
			const info = {
				url: url.pathname,
				rev: url.searchParams.get("rev"),
				ref: url.searchParams.get("ref"),
				dir: url.searchParams.get("dir"),
				submodules: url.searchParams.get("submodules") === "true",
			};

			const root = await nix.evaluate(
				`
				let
					info = builtins.fromJSON ''${JSON.stringify(info)}'';
				in
					builtins.fetchGit (
						{ url = info.url; }
						// (if info.rev != null then { rev = info.rev; } else {})
						// (if info.ref != null then { ref = info.ref; } else {})
						// (if info.submodules != null then { submodules = info.submodules; } else {})
					)
			`,
				{
					impure: true,
					json: true,
				},
			);

			let storePath = await nix.realise(root);

			if (info.dir) {
				storePath = path.join(storePath, info.dir);
			}

			return {
				source: Source.Git,
				...info,
				path: storePath,
			};
		}
		case "github:": {
			const [owner, repo] = url.pathname.split("/");

			const repoResponse = await fetch(
				`https://api.github.com/repos/${owner}/${repo}`,
			);

			if (!repoResponse.ok) {
				log.fatal(
					`GitHub API error: ${repoResponse.status} ${repoResponse.statusText}`,
				);
				process.exit(1);
			}

			const { default_branch } = await repoResponse.json();

			const commitsResponse = await fetch(
				`https://api.github.com/repos/${owner}/${repo}/commits/${default_branch}`,
			);

			if (!commitsResponse.ok) {
				log.fatal(
					`GitHub API error: ${commitsResponse.status} ${commitsResponse.statusText}`,
				);
				process.exit(1);
			}

			const { sha } = await commitsResponse.json();

			const info = {
				owner,
				repo,
				rev: url.searchParams.get("rev") ?? sha,
				dir: url.searchParams.get("dir"),
				host: url.searchParams.get("host") ?? "github.com",
			};

			const root = await nix.evaluate(
				`
				builtins.fetchTarball {
					url = "https://api.${info.host}/repos/${info.owner}/${info.repo}/tarball/${info.rev}";
				}
			`,
				{
					impure: true,
					json: true,
				},
			);

			let storePath = await nix.realise(root);

			if (info.dir) {
				storePath = path.join(storePath, info.dir);
			}

			return {
				source: Source.Github,
				...info,
				path: storePath,
			};
		}
		case "gitlab:": {
			const [owner, repo] = url.pathname.split("/");

			const repoResponse = await fetch(
				`https://${url.searchParams.get("host") ?? "gitlab.com"}/api/v4/projects/${owner}%2F${repo}`,
			);

			if (!repoResponse.ok) {
				log.fatal(
					`GitLab API error: ${repoResponse.status} ${repoResponse.statusText}`,
				);
				process.exit(1);
			}

			const { default_branch } = await repoResponse.json();

			const commitsResponse = await fetch(
				`https://${url.searchParams.get("host") ?? "gitlab.com"}/api/v4/projects/${owner}%2F${repo}/repository/commits/${default_branch}`,
			);

			if (!commitsResponse.ok) {
				log.fatal(
					`GitLab API error: ${commitsResponse.status} ${commitsResponse.statusText}`,
				);
				process.exit(1);
			}

			const { id } = await commitsResponse.json();

			const info = {
				owner,
				repo,
				rev: url.searchParams.get("rev") ?? id,
				dir: url.searchParams.get("dir"),
				host: url.searchParams.get("host") ?? "gitlab.com",
			};

			const root = await nix.evaluate(
				`
				builtins.fetchTarball {
					url = "https://${info.host}/api/v4/projects/${info.owner}%2F${info.repo}/repository/archive.tar.gz?sha=${info.rev}";
				}
			`,
				{
					impure: true,
					json: true,
				},
			);

			let storePath = await nix.realise(root);

			if (info.dir) {
				storePath = path.join(storePath, info.dir);
			}

			return {
				source: Source.Gitlab,
				...info,
				path: storePath,
			};
		}
		case "sourcehut:": {
			const [owner, repo] = url.pathname.split("/");

			log.fatal("Sourcehut support is not yet implemented.");
			process.exit(1);

			return {
				source: Source.Sourcehut,
				owner,
				repo,
				rev: url.searchParams.get("rev"),
				dir: url.searchParams.get("dir"),
			};
		}
		case "tarball:": {
			let tarball = url.pathname;

			if (!tarball.startsWith("http://") && !tarball.startsWith("https://")) {
				tarball = `http://${tarball}`;
			}

			if (!isUrl(tarball)) {
				log.fatal(`Invalid tarball URL: ${tarball}`);
				process.exit(1);
			}

			const root = await nix.evaluate(
				`
				builtins.fetchTarball {
					url = "${tarball}";
				}
			`,
				{
					impure: true,
					json: true,
				},
			);

			const storePath = await nix.realise(root);

			return {
				source: Source.Tarball,
				url: tarball,
				path: storePath,
			};
		}
		case "http:":
		case "https:": {
			const tarball = url.href;

			const root = await nix.evaluate(
				`
				builtins.fetchTarball {
					url = "${tarball}";
				}
			`,
				{
					impure: true,
					json: true,
				},
			);

			const storePath = await nix.realise(root);

			return {
				source: Source.Tarball,
				url: tarball,
				path: storePath,
			};
		}
		default: {
			log.fatal(`Unsupported URI protocol: ${url.protocol}`);
			process.exit(1);
		}
	}
};
